Add PostgreSQL 17 support with pgrx 0.12.7

Fail to build against PostgreSQL 17 and pgrx 0.12.7

```
error[E0277]: the trait bound `Vec<Option<&str>>: ArgAbi<'_>` is not satisfied
   --> src/lib.rs:132:9
    |
125 |     #[pg_extern]
    |     ------------ in this procedural macro expansion
...
132 |         recipients: default!(Option<Vec<Option<&str>>>, "NULL"),
    |         ^^^^^^^^^^ the trait `ArgAbi<'_>` is not implemented for `Vec<Option<&str>>`, which is required by `Option<Vec<Option<&str>>>: ArgAbi<'_>`
    |
    = help: the following other types implement trait `ArgAbi<'fcx>`:
              Vec<Option<T>>
              Vec<T>
              Vec<u8>
    = note: required for `Option<Vec<Option<&str>>>` to implement `ArgAbi<'_>`
note: required by a bound in `pgrx::callconv::Args::<'a, 'fcx>::next_arg_unchecked`
   --> /home/vagrant/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pgrx-0.12.7/src/callconv.rs:931:41
    |
931 |     pub unsafe fn next_arg_unchecked<T: ArgAbi<'fcx>>(&mut self) -> Option<T> {
    |                                         ^^^^^^^^^^^^ required by this bound in `Args::<'a, 'fcx>::next_arg_unchecked`
    = note: this error originates in the attribute macro `pg_extern` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Vec<Option<&str>>: ArgAbi<'_>` is not satisfied
   --> src/lib.rs:133:9
    |
125 |     #[pg_extern]
    |     ------------ in this procedural macro expansion
...
133 |         ccs: default!(Option<Vec<Option<&str>>>, "NULL"),
    |         ^^^ the trait `ArgAbi<'_>` is not implemented for `Vec<Option<&str>>`, which is required by `Option<Vec<Option<&str>>>: ArgAbi<'_>`
    |
    = help: the following other types implement trait `ArgAbi<'fcx>`:
              Vec<Option<T>>
              Vec<T>
              Vec<u8>
    = note: required for `Option<Vec<Option<&str>>>` to implement `ArgAbi<'_>`
note: required by a bound in `pgrx::callconv::Args::<'a, 'fcx>::next_arg_unchecked`
   --> /home/vagrant/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pgrx-0.12.7/src/callconv.rs:931:41
    |
931 |     pub unsafe fn next_arg_unchecked<T: ArgAbi<'fcx>>(&mut self) -> Option<T> {
    |                                         ^^^^^^^^^^^^ required by this bound in `Args::<'a, 'fcx>::next_arg_unchecked`
    = note: this error originates in the attribute macro `pg_extern` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Vec<Option<&str>>: ArgAbi<'_>` is not satisfied
   --> src/lib.rs:134:9
    |
125 |     #[pg_extern]
    |     ------------ in this procedural macro expansion
...
134 |         bccs: default!(Option<Vec<Option<&str>>>, "NULL"),
    |         ^^^^ the trait `ArgAbi<'_>` is not implemented for `Vec<Option<&str>>`, which is required by `Option<Vec<Option<&str>>>: ArgAbi<'_>`
    |
    = help: the following other types implement trait `ArgAbi<'fcx>`:
              Vec<Option<T>>
              Vec<T>
              Vec<u8>
    = note: required for `Option<Vec<Option<&str>>>` to implement `ArgAbi<'_>`
note: required by a bound in `pgrx::callconv::Args::<'a, 'fcx>::next_arg_unchecked`
   --> /home/vagrant/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pgrx-0.12.7/src/callconv.rs:931:41
    |
931 |     pub unsafe fn next_arg_unchecked<T: ArgAbi<'fcx>>(&mut self) -> Option<T> {
    |                                         ^^^^^^^^^^^^ required by this bound in `Args::<'a, 'fcx>::next_arg_unchecked`
    = note: this error originates in the attribute macro `pg_extern` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `pg_smtp_client` (lib) due to 3 previous errors
error: Bad exit status from /var/tmp/rpm-tmp.4WqChe (%build)

```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -18,13 +18,37 @@ on:
 jobs:
   test:
     name: Test
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        PG_VERSION: [pg14, pg15, pg16, pg17]
+    env:
+      PG_VERSION: ${{ matrix.PG_VERSION }}
     steps:
-      - uses: actions/checkout@v4
-      - uses: actions-rust-lang/setup-rust-toolchain@v1
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Rust Toolchain
+        uses: actions-rust-lang/setup-rust-toolchain@v1
         with:
-          toolchain: 1.77.2
+          toolchain: 1.82.0
           components: clippy, rustfmt
-      - run: make init
-      - run: make lint
-      - run: make test
+
+      - name: Cache Rust
+        uses: Swatinem/rust-cache@v2
+
+      - name: Cache pgrx
+        uses: actions/cache@v4
+        with:
+          path: |
+            ~/.pgrx/
+          key: ${{ runner.os }}-${{ matrix.PG_VERSION }}-pgrx
+
+      - name: Init
+        run: make init
+
+      - name: Lint
+        run: make lint
+
+      - name: Test
+        run: make test

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -1,5 +1,10 @@
 # pg_smtp_client
 
+## Install dependencies (needed for testing postgres 17)
+```shell
+sudo apt-get install bison flex
+```
+
 ## Setup
 
 ```shell

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,22 +1,36 @@
 [package]
 name = "pg_smtp_client"
-version = "0.1.1"
+version = "0.2.0"
 edition = "2021"
 publish = false
 
 [lib]
-crate-type = ["cdylib"]
+crate-type = ["cdylib", "lib"]
+
+[[bin]]
+name = "pgrx_embed_pg_smtp_client"
+path = "./src/bin/pgrx_embed.rs"
 
 [features]
-default = ["pg16"]
+default = ["pg17"]
 pg14 = ["pgrx/pg14", "pgrx-tests/pg14" ]
 pg15 = ["pgrx/pg15", "pgrx-tests/pg15" ]
 pg16 = ["pgrx/pg16", "pgrx-tests/pg16" ]
+pg17 = ["pgrx/pg17", "pgrx-tests/pg17" ]
 pg_test = []
 
 [dependencies]
-pgrx = "=0.11.4"
-lettre = "=0.11.7"
+pgrx = "0.12.7"
+lettre = "=0.11.10"
 
 [dev-dependencies]
-pgrx-tests = "=0.11.4"
+pgrx-tests = "0.12.7"
+
+[profile.dev]
+panic = "unwind"
+
+[profile.release]
+panic = "unwind"
+opt-level = 3
+lto = "fat"
+codegen-units = 1

--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,10 @@
 PYTHON ?= $(shell which python3 || which python)
 PIP ?= $(shell which pip3 || which pip)
-PG_VERSION ?= 16
+PG_VERSION ?= pg17
+
+.PHONY: build
+build:
+	@cargo build
 
 .PHONY: run
 run:
@@ -12,13 +16,13 @@ clean:
 	
 .PHONY: init
 init:
-	@cargo install --locked cargo-pgrx@0.11.4
-	@if ! cargo pgrx info version pg$(PG_VERSION) >/dev/null 2>&1; then cargo pgrx init --pg$(PG_VERSION) download; else echo "pg$(PG_VERSION) already installed"; fi
+	@cargo install --locked cargo-pgrx@0.12.7
+	@if ! cargo pgrx info version $(PG_VERSION) >/dev/null 2>&1; then cargo pgrx init --$(PG_VERSION) download; else echo "$(PG_VERSION) already installed"; fi
 
 .PHONY: lint
 lint:
 	@cargo fmt --check
-	@cargo clippy
+	@cargo clippy --no-default-features --features $(PG_VERSION)
 
 .PHONY: test
 test:
@@ -27,4 +31,4 @@ test:
 	@$(PIP) install aiosmtpd
 	@trap 'kill `cat /tmp/smtpd.pid`' EXIT; \
 	$(PYTHON) -m aiosmtpd -n & echo $$! > /tmp/smtpd.pid; \
-	cargo pgrx test pg$(PG_VERSION)
+	cargo pgrx test

--- a/README.md
+++ b/README.md
@@ -1,5 +1,7 @@
 # pg_smtp_client
 
+[![CI](https://github.com/brianpursley/pg_smtp_client/actions/workflows/ci.yml/badge.svg)](https://github.com/brianpursley/pg_smtp_client/actions/workflows/ci.yml)
+
 A Postgres extension to send emails using SMTP.
 
 ## Installation

--- a/src/bin/pgrx_embed.rs
+++ b/src/bin/pgrx_embed.rs
@@ -1,0 +1,1 @@
+::pgrx::pgrx_embed!();

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -15,12 +15,12 @@ mod smtp_client {
     use lettre::transport::smtp::client::{Tls, TlsParameters};
     use lettre::{Message, SmtpTransport, Transport};
 
-    fn create_mailer(
-        smtp_server: Option<&str>,
+    fn create_mailer<'a>(
+        smtp_server: Option<&'a str>,
         smtp_port: Option<i32>,
         smtp_tls: Option<bool>,
-        smtp_username: Option<&str>,
-        smtp_password: Option<&str>,
+        smtp_username: Option<&'a str>,
+        smtp_password: Option<&'a str>,
     ) -> Result<SmtpTransport, String> {
         let server = smtp_server
             .map_or_else(guc::get_smtp_server, |x| Some(x.to_string()))
@@ -57,14 +57,14 @@ mod smtp_client {
     }
 
     #[allow(clippy::too_many_arguments)]
-    fn create_message(
-        subject: &str,
-        body: &str,
+    fn create_message<'a>(
+        subject: &'a str,
+        body: &'a str,
         html: bool,
-        from_address: Option<&str>,
-        recipients: Option<Vec<Option<&str>>>,
-        ccs: Option<Vec<Option<&str>>>,
-        bccs: Option<Vec<Option<&str>>>,
+        from_address: Option<&'a str>,
+        recipients: Option<Vec<Option<String>>>,
+        ccs: Option<Vec<Option<String>>>,
+        bccs: Option<Vec<Option<String>>>,
         keep_bcc_header: bool,
     ) -> Result<Message, String> {
         let mut email = Message::builder().subject(subject);
@@ -83,7 +83,7 @@ mod smtp_client {
         }
 
         if let Some(items) = recipients {
-            for addr in items.into_iter().flatten() {
+            for addr in items.iter().flatten() {
                 if let Ok(parsed_addr) = addr.parse() {
                     email = email.to(parsed_addr);
                 } else {
@@ -93,7 +93,7 @@ mod smtp_client {
         }
 
         if let Some(items) = ccs {
-            for addr in items.into_iter().flatten() {
+            for addr in items.iter().flatten() {
                 if let Ok(parsed_addr) = addr.parse() {
                     email = email.cc(parsed_addr);
                 } else {
@@ -106,7 +106,7 @@ mod smtp_client {
             if keep_bcc_header {
                 email = email.keep_bcc();
             }
-            for addr in items.into_iter().flatten() {
+            for addr in items.iter().flatten() {
                 if let Ok(parsed_addr) = addr.parse() {
                     email = email.bcc(parsed_addr);
                 } else {
@@ -124,19 +124,19 @@ mod smtp_client {
 
     #[pg_extern]
     #[allow(clippy::too_many_arguments)]
-    fn send_email(
-        subject: &str,
-        body: &str,
+    fn send_email<'a>(
+        subject: &'a str,
+        body: &'a str,
         html: default!(bool, "false"),
-        from_address: default!(Option<&str>, "NULL"),
-        recipients: default!(Option<Vec<Option<&str>>>, "NULL"),
-        ccs: default!(Option<Vec<Option<&str>>>, "NULL"),
-        bccs: default!(Option<Vec<Option<&str>>>, "NULL"),
-        smtp_server: default!(Option<&str>, "NULL"),
+        from_address: default!(Option<&'a str>, "NULL"),
+        recipients: default!(Option<Vec<Option<String>>>, "NULL"),
+        ccs: default!(Option<Vec<Option<String>>>, "NULL"),
+        bccs: default!(Option<Vec<Option<String>>>, "NULL"),
+        smtp_server: default!(Option<&'a str>, "NULL"),
         smtp_port: default!(Option<i32>, "NULL"),
         smtp_tls: default!(Option<bool>, "NULL"),
-        smtp_username: default!(Option<&str>, "NULL"),
-        smtp_password: default!(Option<&str>, "NULL"),
+        smtp_username: default!(Option<&'a str>, "NULL"),
+        smtp_password: default!(Option<&'a str>, "NULL"),
     ) -> String {
         let mailer = create_mailer(
             smtp_server,
@@ -224,7 +224,7 @@ mod smtp_client {
                 "test body",
                 false,
                 Some("from@example.com"),
-                Some(vec![Some("to@example.com")]),
+                Some(vec![Some("to@example.com".to_string())]),
                 None,
                 None,
                 Some("127.0.0.1"),
@@ -251,7 +251,7 @@ mod smtp_client {
                 "test body",
                 false,
                 None,
-                Some(vec![Some("to@example.com")]),
+                Some(vec![Some("to@example.com".to_string())]),
                 None,
                 None,
                 None,
@@ -271,7 +271,7 @@ mod smtp_client {
                 "test body",
                 false,
                 None,
-                Some(vec![Some("to@example.com")]),
+                Some(vec![Some("to@example.com".to_string())]),
                 None,
                 None,
                 None,
@@ -289,7 +289,7 @@ mod smtp_client {
                 "Test Body",
                 false,
                 Some("from@example.com"),
-                Some(vec![Some("to@example.com")]),
+                Some(vec![Some("to@example.com".to_string())]),
                 None,
                 None,
                 false,
@@ -310,7 +310,10 @@ mod smtp_client {
                 "Test Body",
                 false,
                 Some("from@example.com"),
-                Some(vec![Some("to1@example.com"), Some("to2@example.com")]),
+                Some(vec![
+                    Some("to1@example.com".to_string()),
+                    Some("to2@example.com".to_string()),
+                ]),
                 None,
                 None,
                 false,
@@ -331,9 +334,9 @@ mod smtp_client {
                 "Test Body",
                 false,
                 Some("from@example.com"),
-                Some(vec![Some("to@example.com")]),
-                Some(vec![Some("cc@example.com")]),
-                Some(vec![Some("bcc@example.com")]),
+                Some(vec![Some("to@example.com".to_string())]),
+                Some(vec![Some("cc@example.com".to_string())]),
+                Some(vec![Some("bcc@example.com".to_string())]),
                 true,
             )
             .unwrap();
@@ -352,9 +355,18 @@ mod smtp_client {
                 "Test Body",
                 false,
                 Some("from@example.com"),
-                Some(vec![Some("to1@example.com"), Some("to2@example.com")]),
-                Some(vec![Some("cc1@example.com"), Some("cc2@example.com")]),
-                Some(vec![Some("bcc1@example.com"), Some("bcc2@example.com")]),
+                Some(vec![
+                    Some("to1@example.com".to_string()),
+                    Some("to2@example.com".to_string()),
+                ]),
+                Some(vec![
+                    Some("cc1@example.com".to_string()),
+                    Some("cc2@example.com".to_string()),
+                ]),
+                Some(vec![
+                    Some("bcc1@example.com".to_string()),
+                    Some("bcc2@example.com".to_string()),
+                ]),
                 true,
             )
             .unwrap();
@@ -376,7 +388,7 @@ mod smtp_client {
                 "Test Body",
                 false,
                 None,
-                Some(vec![Some("to@example.com")]),
+                Some(vec![Some("to@example.com".to_string())]),
                 None,
                 None,
                 false,
@@ -400,7 +412,7 @@ mod smtp_client {
                 "Test Body",
                 false,
                 Some("override@example.com"),
-                Some(vec![Some("to@example.com")]),
+                Some(vec![Some("to@example.com".to_string())]),
                 None,
                 None,
                 false,
@@ -421,7 +433,7 @@ mod smtp_client {
                 "Test Body",
                 false,
                 None,
-                Some(vec![Some("to@example.com")]),
+                Some(vec![Some("to@example.com".to_string())]),
                 None,
                 None,
                 false,
@@ -430,7 +442,7 @@ mod smtp_client {
             assert!(result.is_err());
             assert_eq!(
                 result.unwrap_err(),
-                format!("From address not provided and no default configured")
+                "From address not provided and no default configured"
             );
         }
     }

